---
title: "Statistical significance tests for DMLs and DMRs"
author: "Cemile Balkas"
date: "2025-06-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!requireNamespace("BiocManager", quitetly = TRUE)) {
      install.packages("BiocManager")
}
if (!requireNamespace("GenomeInfoDb", quietly = TRUE)) {
  BiocManager::install("GenomeInfoDb")
}
if (!requireNamespace("GenomicRanges", quietly = TRUE)) {
  BiocManager::install("GenomicRanges")
}
if (!requireNamespace("GenomicFeatures", quietly = TRUE)) {
  BiocManager::install("GenomicFeatures")
}
library(dplyr)
library(GenomeInfoDb)
library(GenomicRanges)
library(GenomicFeatures)
```

## Get annotation file from Ensembl (Bash)

```{bash get-annotation}
# More info on annotation
# https://www.ensembl.org/info/data/ftp/index.html
# https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/

wget ftp://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/Homo_sapiens.GRCh38.114.gtf.gz -P data/annotation_data

gunzip data/annotation_data/Homo_sapiens.GRCh38.114.gtf.gz
```

## Load annotation file from Ensembl

```{r load-annotation}
# Import ensembl annotation as GRanges -------------------------------
ensembl_annot <- rtracklayer::import("data/annotation_data/Homo_sapiens.GRCh38.114.gtf")
genome(ensembl_annot) <- "hg38"

# Change the chromosome names to UCSC style (chr1, chr2...chrM) adapted from: https://www.bioconductor.org/packages/release/bioc/vignettes/GenomeInfoDb/inst/doc/GenomeInfoDb.pdf
sequences <- seqlevels(ensembl_annot)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_annot <- renameSeqlevels(ensembl_annot, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_annot <- keepSeqlevels(ensembl_annot, auto, pruning.mode = "coarse")

# Import ensembl annotation as TxDb -------------------------------
# Metadata for TxDb
ensembl_metadata <- data.frame(
  name = c("Genome", "Resource URL", "Type of Gene ID", "Full dataset"),
  value = c(
    "hg38",
    "https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/",
    "Ensembl Gene ID",
    "yes"
  ),
  stringsAsFactors = FALSE
)

# Make ensembl TxDb 
ensembl_txdb <- txdbmaker::makeTxDbFromGFF(
  file = "data/annotation_data/Homo_sapiens.GRCh38.114.gtf",
  format = "gtf",
  dataSource = "Ensembl",
  organism = "Homo sapiens",
  taxonomyId = 9606,
  metadata = ensembl_metadata
)

# Change the chromosome names to UCSC style (chr1, chr2...chrM)
sequences <- seqlevels(ensembl_txdb)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_txdb <- renameSeqlevels(ensembl_txdb, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_txdb <- keepSeqlevels(ensembl_txdb, auto)

```

## Explore the annotation objects

```{r explore-annotation-GRange}
# Validation checks between objects ------------
# following gives the same number of transcript for both object as a validation check
table(ensembl_annot@elementMetadata@listData[["type"]])
length(transcripts(ensembl_txdb))

# Check how many features are associated with each chromosome
table(seqnames(ensembl_annot))

# Check the total number of features there are in our genome
summary(table(seqnames(ensembl_annot)))

# indexing ensembl GRange
head(table(ensembl_annot@elementMetadata$gene_id)) # first row gene names, second row gene features (ie. exons, introns)

head(table(ensembl_annot@elementMetadata$gene_name))
```

```{r GRange-index-into-data.frame}
ensembl_gr_df <- data.frame("gene_id" = ensembl_annot@elementMetadata$gene_id,
                           "gene_name" = ensembl_annot@elementMetadata$gene_name,
                           "gene_biotype" = ensembl_annot@elementMetadata$gene_biotype,
                           "type" = ensembl_annot@elementMetadata$type,
                           "start" = ensembl_annot@ranges@start,
                           "width" = ensembl_annot@ranges@width,
                           "chr" = ensembl_annot@seqnames)
```

## Subset and filter the annotation

```{r mRNA-gene-width}

mrna_df <- filter(ensembl_gr_df, gene_biotype == "protein_coding", type == "gene")
hist(log10(mrna_df$width), breaks = 60)

mean(mrna_df$width)

# this is for gene body length for protein coding genes
```
## What annotations have width of 1nt?

```{r}
# use indexing to set a value of 1 and making a dataframe
width_1_items <- ensembl_annot[width(ensembl_annot) == 1] %>%
  as.data.frame()

table(width_1_items$type)

# exons that are just 1bp long in the annotation, weird?? Can look in genome browser
weird_exons <- width_1_items %>%
  filter(type == "exon")

# Filter for 1bp exons directly from the GRanges
weird_exons2 <- ensembl_annot[width(ensembl_annot) == 1 & ensembl_annot$type == "exon"]
mcols(weird_exons2) <- NULL

# Export to BED to see in igv
rtracklayer::export(weird_exons2, "weird_exons.bed", format = "BED")

```

## Define gene TSS

```{r}
# using the promoters function (from GenomicRanges package): it defines the tss and input bases up and downstream
tss <- promoters(ensembl_annot[ensembl_annot$type == "gene"], upstream = 0, downstream = 0) %>%
  as.data.frame()
nrow(tss)

# 75,022 TSS, check if that is the same # of genes:
summary(ensembl_annot[ensembl_annot$type == "gene"])
# yep...

```

```{r shortest-gene-in-human-genome}

ensembl_df_genes <- ensembl_annot %>%
  as.data.frame %>%
  filter(type == "gene")

ensembl_df_genes %>%
  arrange(ensembl_df_genes$width) %>%
  head()

```
## What percentage of the genome are exons?

```{r}
# first make an object of exons and reduce them
total_exon_length <- ensembl_annot[ensembl_annot$type == "exon"] %>%
  # the reduce here takes the union of overlapping exons
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

total_exon_length/3.2e9 * 100
# Our filtered genome without the non autosomal chromosomes is about 5.7% exons

# Plot the range of exon sizes we can simply remove the sum function and plot
exon_lengths <- ensembl_annot[ensembl_annot$type == "exon"] %>%
  GenomicRanges::reduce()

# plotting
hist(width(exon_lengths), breaks = 100, xlim = c(0,10000))

# average exon length
average_exon_length <- mean(width(exon_lengths))
average_exon_length 
# it is 453nt
```

## What percentage of the genome is compriesed of gene bodies?

```{r}
# same thing as exons but change index in type to gene
reduced_gene_bodies <- ensembl_annot[ensembl_annot$type == "gene"] %>%
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

reduced_gene_bodies/3.2e9 * 100
# ~70% of the genome is made up of gene bodies...

gene_body_lengths <- ensembl_annot[ensembl_annot$type == "gene"] %>%
  GenomicRanges::reduce()
  
hist(width(gene_body_lengths), breaks = 100, xlim = c(0,500000))

# What is average length of a gene body?

mean(width(gene_body_lengths))
# 38Kb

```
## Import Methylation files

```{r}
# Import Methylation loci and DMLs as GRanges -------------------------------
all_dmls_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs.bed")
genome(all_dmls_gr) <- "hg38"
dmls_hyper_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs_HYPER.bed")
genome(dmls_hyper_gr) <- "hg38"
dmls_hypo_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs_HYPO.bed")
genome(dmls_hypo_gr) <- "hg38"
all_cpg_loci <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_CpG_Loci.bed")
genome(all_cpg_loci) <- "hg38"
```

## Define promoters and other genomic regions
```{r}
# promoter function uses the first base in a genbody at the TSS
#add 1Kb upstream and downstream from the TSS to define "promoters"
ensembl_promoters <- promoters(ensembl_annot[ensembl_annot$type == "gene"],
                               upstream = 1e3, 
                               downstream = 1e3)
length(ensembl_promoters)
immediate_downstream <- flank(ensembl_annot[ensembl_annot$type == "gene"], width = 1000, start = FALSE)
length(immediate_downstream)
five_utrs <- ensembl_annot[ensembl_annot$type == "five_prime_utr"]
length(five_utrs)
three_utrs <- ensembl_annot[ensembl_annot$type == "three_prime_utr"]
length(three_utrs)
exons <- ensembl_annot[ensembl_annot$type == "exon"]
length(exons)
introns <- GenomicFeatures::intronsByTranscript(ensembl_txdb, use.names = TRUE) %>% unlist()
length(introns)

exons_from_txdb <- GenomicFeatures::exonsBy(ensembl_txdb, by=c("tx", "gene"), use.names = TRUE) %>% unlist()
length(exons_from_txdb) # this was just to check the previous exon calculation and they are the same so np

# Create region list
region_list <- list(
  Promoters = ensembl_promoters,
  immediateDownstream = immediate_downstream,
  fiveUTRs = five_utrs,
  threeUTRs = three_utrs,
  Exons = exons,
  Introns = introns
)

```
 
## How many of DMLs overlap with promoters
```{r}
# Query file(dmls_hyper_gr) and subject file (ensembl_promoters)
promoter_overlaps_hyper_dmls <- findOverlaps(dmls_hyper_gr, ensembl_promoters)
promoter_overlaps_hyper_dmls

promoter_overlaps_hypo_dmls <- findOverlaps(dmls_hypo_gr, ensembl_promoters)
promoter_overlaps_hypo_dmls

promoter_overlaps_dmls <- findOverlaps(all_dmls_gr, ensembl_promoters)

# Count DMLs overlapping promoters
dml_in_promoter <- sum(countOverlaps(all_dmls_gr, ensembl_promoters) > 0)
cpg_in_promoter <- sum(countOverlaps(all_cpg_loci, ensembl_promoters) > 0)

# Total DMLs and CpGs
total_dmls <- length(all_dmls_gr)
total_cpgs <- length(all_cpg_loci)

# Not in promoter
dml_not_in_promoter <- total_dmls - dml_in_promoter
cpg_not_in_promoter <- total_cpgs - cpg_in_promoter

expected_dmls_in_promoter <- total_dmls * (cpg_in_promoter / total_cpgs)

promoter_contingency_table <- matrix(
  c(dml_in_promoter, dml_not_in_promoter,
    cpg_in_promoter, cpg_not_in_promoter),
  nrow = 2,
  byrow = TRUE,
  dimnames = list(
    c("DMLs", "CpGs"),
    c("In_Promoter", "Not_In_Promoter")
  )
)

fisher_result <- fisher.test(promoter_contingency_table)

print(promoter_contingency_table)
cat("\nExpected DMLs in promoter:", round(expected_dmls_in_promoter), "\n")
cat("Observed DMLs in promoter:", dml_in_promoter, "\n")
cat("Odds ratio:", fisher_result$estimate, "\n")
cat("P-value:", format(fisher_result$p.value, scientific = TRUE), "\n")

```

# Promoter test using chippeakanno
```{r}
chipanno_dml <- assignChromosomeRegion(
  all_dmls_gr,
  TxDb = ensembl_txdb,
  nucleotideLevel = FALSE,
  proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000),
  precedence = c("Promoters", "Exons", "Introns")
)

chipanno_cpg <- assignChromosomeRegion(
  all_cpg_loci,
  TxDb = ensembl_txdb,
  nucleotideLevel = FALSE,
  proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000),
  precedence = c("Promoters", "Exons", "Introns")
)

dml_promoters <- round(chipanno_dml$percentage["Promoters"] / 100 * length(all_dmls_gr))
cpg_promoters <- round(chipanno_cpg$percentage["Promoters"] / 100 * length(all_cpg_loci))

dml_non_promoters <- length(all_dmls_gr) - dml_promoters
cpg_non_promoters <- length(all_cpg_loci) - cpg_promoters

contingency_chipanno <- matrix(
  c(dml_promoters, dml_non_promoters,
    cpg_promoters, cpg_non_promoters),
  nrow = 2,
  byrow = TRUE,
  dimnames = list(c("DMLs", "CpGs"), c("In_Promoter", "Not_In_Promoter"))
)

fisher_chipanno <- fisher.test(contingency_chipanno)

print(contingency_chipanno)
cat("Expected DMLs in promoter:", round(length(all_dmls_gr) * (cpg_promoters / length(all_cpg_loci))), "\n")
cat("Observed DMLs in promoter:", dml_promoters, "\n")
cat("Odds ratio:", fisher_chipanno$estimate, "\n")
cat("P-value:", format(fisher_chipanno$p.value, scientific = TRUE), "\n")
```


## Count Overlaps with DMLs and all CpGs

```{r}
assign_regions_exclusively <- function(gr_target, region_list) {
  labels <- rep("Intergenic.Region", length(gr_target))
  assigned <- rep(FALSE, length(gr_target))
  
  for (region_name in names(region_list)) {
    overlaps <- findOverlaps(gr_target[!assigned], region_list[[region_name]])
    hits <- queryHits(overlaps)
    # Assign region
    labels[which(!assigned)[hits]] <- region_name
    assigned[which(!assigned)[hits]] <- TRUE
  }
  
  table(factor(labels, levels = c(names(region_list), "Intergenic.Region")))
}

# Use the fixed exclusive assignment function
dml_counts <- assign_regions_exclusively(all_dmls_gr, region_list)
cpg_counts <- assign_regions_exclusively(all_cpg_loci, region_list)

# Combine into data frame
counts_df <- data.frame(
  feature = names(dml_counts),
  DMLs = as.numeric(dml_counts),
  CpGs = as.numeric(cpg_counts)
)

```
## Fisher's Exact Test
```{r}

total_dmls <- sum(counts_df$DMLs) 
total_cpgs <- sum(counts_df$CpGs)

library(dplyr)

fisher_results <- counts_df %>%
  rowwise() %>%
  mutate(
    not_dmls = total_dmls - DMLs,
    not_cpgs = total_cpgs - CpGs,
    p_value = fisher.test(matrix(c(DMLs, not_dmls, CpGs, not_cpgs), nrow = 2))$p.value,
    odds_ratio = fisher.test(matrix(c(DMLs, not_dmls, CpGs, not_cpgs), nrow = 2))$estimate
  ) %>%
  ungroup() %>%
  mutate(
    p_adj = p.adjust(p_value, method = "BH")
  )

fisher_results <- fisher_results %>%
  mutate(
    p_value_sci = format(p_value, scientific = TRUE),
    p_adj_sci = format(p_adj, scientific = TRUE)
  )


```
## ChipPeakAnno  functions
```{r}
chipanno_dml <- assignChromosomeRegion(
  all_dmls_gr,
  TxDb = ensembl_txdb,
  nucleotideLevel = FALSE,
  proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000),
  immediate.downstream.cutoff = c(upstream = 0, downstream = 1000)
  # precedence = c("Promoters", "immediateDownstream", "fiveUTRs", "threeUTRs", "Exons", "Introns")
)

chipanno_cpg <- assignChromosomeRegion(
  all_cpg_loci,
  TxDb = ensembl_txdb,
  nucleotideLevel = FALSE,
  proximal.promoter.cutoff = c(upstream = 1000, downstream = 1000),
  immediate.downstream.cutoff = c(upstream = 0, downstream = 1000)
  # precedence = c("Promoters", "immediateDownstream", "fiveUTRs", "threeUTRs", "Exons", "Introns")
)

dml_percent <- chipanno_dml$percentage
cpg_percent <- chipanno_cpg$percentage

# Convert to counts
dml_counts2 <- round(dml_percent / 100 * length(all_dmls_gr))
cpg_counts2 <- round(cpg_percent / 100 * length(all_cpg_loci))

# Convert to data frame

counts_df_chipanno2 <- data.frame(
  feature = names(dml_counts),
  DMLs = as.numeric(dml_counts2),
  CpGs = as.numeric(cpg_counts2)
)



```

```{r}
library(dplyr)

total_dmls <- sum(counts_df_chipanno$DMLs)
total_cpgs <- sum(counts_df_chipanno$CpGs)

fisher_chipanno <- counts_df_chipanno %>%
  rowwise() %>%
  mutate(
    not_dmls = total_dmls - DMLs,
    not_cpgs = total_cpgs - CpGs,
    p_value = fisher.test(matrix(c(DMLs, not_dmls, CpGs, not_cpgs), nrow = 2))$p.value,
    odds_ratio = fisher.test(matrix(c(DMLs, not_dmls, CpGs, not_cpgs), nrow = 2))$estimate
  ) %>%
  ungroup() %>%
  mutate(p_adj = p.adjust(p_value, method = "BH"))

```


## Visual represtentation of CpGs

All CpG sites
All DMLs
Hyper DMLs
Hypo DMLs


## Understand how each of the chromosome regions are counted in ChipPeakAnno


## Calculate expected values

For each type of annotation (e.g. promoter regions, intragenic region etc ) check how many dmls do we expect to land on each of those by chance


## Perform fishers exact test for DMLs
