---
title: "Statistical significance tests for DMLs and DMRs"
author: "Cemile Balkas"
date: "2025-06-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!requireNamespace("BiocManager", quitetly = TRUE)) {
      install.packages("BiocManager")
}
if (!requireNamespace("GenomeInfoDb", quietly = TRUE)) {
  BiocManager::install("GenomeInfoDb")
}
if (!requireNamespace("GenomicRanges", quietly = TRUE)) {
  BiocManager::install("GenomicRanges")
}
if (!requireNamespace("GenomicFeatures", quietly = TRUE)) {
  BiocManager::install("GenomicFeatures")
}
library(dplyr)
library(GenomeInfoDb)
library(GenomicRanges)
library(GenomicFeatures)
```

## Get annotation file from Ensembl (Bash)

```{bash get-annotation}
# More info on annotation
# https://www.ensembl.org/info/data/ftp/index.html
# https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/

wget ftp://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/Homo_sapiens.GRCh38.114.gtf.gz -P data/annotation_data

gunzip data/annotation_data/Homo_sapiens.GRCh38.114.gtf.gz
```

## Load annotation file from Ensembl

```{r load-annotation}
# Import ensembl annotation as GRanges -------------------------------
ensembl_annot <- rtracklayer::import("data/annotation_data/Homo_sapiens.GRCh38.114.gtf")
genome(ensembl_annot) <- "hg38"

# Change the chromosome names to UCSC style (chr1, chr2...chrM) adapted from: https://www.bioconductor.org/packages/release/bioc/vignettes/GenomeInfoDb/inst/doc/GenomeInfoDb.pdf
sequences <- seqlevels(ensembl_annot)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_annot <- renameSeqlevels(ensembl_annot, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_annot <- keepSeqlevels(ensembl_annot, auto, pruning.mode = "coarse")

# Import ensembl annotation as TxDb -------------------------------
# Metadata for TxDb
ensembl_metadata <- data.frame(
  name = c("Genome", "Resource URL", "Type of Gene ID", "Full dataset"),
  value = c(
    "hg38",
    "https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/",
    "Ensembl Gene ID",
    "yes"
  ),
  stringsAsFactors = FALSE
)

# Make ensembl TxDb 
ensembl_txdb <- txdbmaker::makeTxDbFromGFF(
  file = "data/annotation_data/Homo_sapiens.GRCh38.114.gtf",
  format = "gtf",
  dataSource = "Ensembl",
  organism = "Homo sapiens",
  taxonomyId = 9606,
  metadata = ensembl_metadata
)

# Change the chromosome names to UCSC style (chr1, chr2...chrM)
sequences <- seqlevels(ensembl_txdb)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_txdb <- renameSeqlevels(ensembl_txdb, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_txdb <- keepSeqlevels(ensembl_txdb, auto)

```

## Explore the annotation objects

```{r explore-annotation-GRange}
# Validation checks between objects ------------
# following gives the same number of transcript for both object as a validation check
table(ensembl_annot@elementMetadata@listData[["type"]])
length(transcripts(ensembl_txdb))

# Check how many features are associated with each chromosome
table(seqnames(ensembl_annot))

# Check the total number of features there are in our genome
summary(table(seqnames(ensembl_annot)))

# indexing ensembl GRange
head(table(ensembl_annot@elementMetadata$gene_id)) # first row gene names, second row gene features (ie. exons, introns)

head(table(ensembl_annot@elementMetadata$gene_name))
```

```{r GRange-index-into-data.frame}
ensembl_gr_df <- data.frame("gene_id" = ensembl_annot@elementMetadata$gene_id,
                           "gene_name" = ensembl_annot@elementMetadata$gene_name,
                           "gene_biotype" = ensembl_annot@elementMetadata$gene_biotype,
                           "type" = ensembl_annot@elementMetadata$type,
                           "start" = ensembl_annot@ranges@start,
                           "width" = ensembl_annot@ranges@width,
                           "chr" = ensembl_annot@seqnames)
```

## Subset and filter the annotation

```{r}

mrna_df <- filter(ensembl_gr_df, gene_biotype == "protein_coding", type == "gene")
hist(log10(mrna_df$width), breaks = 60)

```

## What annotations have width of 1nt?

```{r}
width_1_items <- ensembl_annot[width(ensembl_annot) == 1] %>%
  as.data.frame()

table(width_1_items$type)

# exons that are just 1bp long in the annotation, weird?? Can look in genome browser
weird_exons <- width_1_items %>%
  filter(type == "exon")

```

## Define gene TSS

```{r}
# using the promoters function: it defines the tss and input bases up and downstream

tss <- promoters(ensembl_annot[ensembl_annot$type == "gene"], upstream = 0, downstream = 0) %>%
  as.data.frame()
nrow(tss)

# 75,022 TSS, check if that is the same # of genes:
summary(ensembl_annot[ensembl_annot$type == "gene"])
# yep...

```


## Visual represtentation of CpGs

All CpG sites
All DMLs
Hyper DMLs
Hypo DMLs


## Understand how each of the chromosome regions are counted in ChipPeakAnno


## Calculate expected values

For each type of annotation (e.g. promoter regions, intragenic region etc ) check how many dmls do we expect to land on each of those by chance


## Perform fishers exact test for DMLs
