---
title: "Statistical significance tests for DMLs and DMRs"
author: "Cemile Balkas"
date: "2025-06-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!requireNamespace("BiocManager", quitetly = TRUE)) {
      install.packages("BiocManager")
}
if (!requireNamespace("GenomeInfoDb", quietly = TRUE)) {
  BiocManager::install("GenomeInfoDb")
}
if (!requireNamespace("GenomicRanges", quietly = TRUE)) {
  BiocManager::install("GenomicRanges")
}
if (!requireNamespace("GenomicFeatures", quietly = TRUE)) {
  BiocManager::install("GenomicFeatures")
}

library(rtracklayer)
library(dplyr)
library(GenomeInfoDb)
library(GenomicRanges)
library(GenomicFeatures)
source("assignChromRegionCounts.R")

```

## Get annotation file from Ensembl (Bash)

```{bash get-annotation}
# More info on annotation
# https://www.ensembl.org/info/data/ftp/index.html
# https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/

wget ftp://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/Homo_sapiens.GRCh38.114.gtf.gz -P data/annotation_data

gunzip data/annotation_data/Homo_sapiens.GRCh38.114.gtf.gz
```

## Load annotation file from Ensembl

```{r load-annotation}
# Import ensembl annotation as GRanges -------------------------------
ensembl_annot <- rtracklayer::import("data/annotation_data/Homo_sapiens.GRCh38.114.gtf")
genome(ensembl_annot) <- "hg38"

# Change the chromosome names to UCSC style (chr1, chr2...chrM) adapted from: https://www.bioconductor.org/packages/release/bioc/vignettes/GenomeInfoDb/inst/doc/GenomeInfoDb.pdf
sequences <- seqlevels(ensembl_annot)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_annot <- renameSeqlevels(ensembl_annot, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_annot <- keepSeqlevels(ensembl_annot, auto, pruning.mode = "coarse")

# Import ensembl annotation as TxDb -------------------------------
# Metadata for TxDb
ensembl_metadata <- data.frame(
  name = c("Genome", "Resource URL", "Type of Gene ID", "Full dataset"),
  value = c(
    "hg38",
    "https://ftp.ensembl.org/pub/release-114/gtf/homo_sapiens/",
    "Ensembl Gene ID",
    "yes"
  ),
  stringsAsFactors = FALSE
)

# Make ensembl TxDb 
ensembl_txdb <- txdbmaker::makeTxDbFromGFF(
  file = "data/annotation_data/Homo_sapiens.GRCh38.114.gtf",
  format = "gtf",
  dataSource = "Ensembl",
  organism = "Homo sapiens",
  taxonomyId = 9606,
  metadata = ensembl_metadata
)

# Change the chromosome names to UCSC style (chr1, chr2...chrM)
sequences <- seqlevels(ensembl_txdb)
ucsc_style <- mapSeqlevels(sequences, "UCSC")
ucsc_style <- ucsc_style[complete.cases(ucsc_style)]
ensembl_txdb <- renameSeqlevels(ensembl_txdb, ucsc_style)

# Filter to autosomal chromosomes
auto <- extractSeqlevelsByGroup(species="Homo sapiens", style="UCSC", group="auto")
ensembl_txdb <- keepSeqlevels(ensembl_txdb, auto)

```

## Explore the annotation objects

```{r explore-annotation-GRange}
# Validation checks between objects ------------
# following gives the same number of transcript for both gr and txdb object as a validation check
table(ensembl_annot@elementMetadata@listData[["type"]])
length(transcripts(ensembl_txdb))

# Check how many features are associated with each chromosome
table(seqnames(ensembl_annot))

# Check the total number of features there are in our genome
summary(table(seqnames(ensembl_annot)))

# indexing ensembl GRange
head(table(ensembl_annot@elementMetadata$gene_id)) # first row gene names, second row gene features (ie. exons, introns)

head(table(ensembl_annot@elementMetadata$gene_name))
```

```{r GRange-index-into-data.frame}
ensembl_gr_df <- data.frame("gene_id" = ensembl_annot@elementMetadata$gene_id,
                           "gene_name" = ensembl_annot@elementMetadata$gene_name,
                           "gene_biotype" = ensembl_annot@elementMetadata$gene_biotype,
                           "type" = ensembl_annot@elementMetadata$type,
                           "start" = ensembl_annot@ranges@start,
                           "width" = ensembl_annot@ranges@width,
                           "chr" = ensembl_annot@seqnames)
```

## Subset and filter the annotation

```{r mRNA-gene-width}

mrna_df <- filter(ensembl_gr_df, gene_biotype == "protein_coding", type == "gene")
hist(log10(mrna_df$width), breaks = 60)

mean(mrna_df$width)

# this is for gene body length for protein coding genes
```
## What annotations have width of 1nt?

```{r}
# use indexing to set a value of 1 and making a dataframe
width_1_items <- ensembl_annot[width(ensembl_annot) == 1] %>%
  as.data.frame()

table(width_1_items$type)

# exons that are just 1bp long in the annotation, weird?? Can look in genome browser
weird_exons <- width_1_items %>%
  filter(type == "exon")

# Filter for 1bp exons directly from the GRanges
weird_exons2 <- ensembl_annot[width(ensembl_annot) == 1 & ensembl_annot$type == "exon"]
mcols(weird_exons2) <- NULL

# Export to BED to see in igv
rtracklayer::export(weird_exons2, "weird_exons.bed", format = "BED")

```

## Define gene TSS

```{r}
# using the promoters function (from GenomicRanges package): it defines the tss and input bases up and downstream
tss <- promoters(ensembl_annot[ensembl_annot$type == "gene"], upstream = 0, downstream = 0) %>%
  as.data.frame()
nrow(tss)

# 75,022 TSS, check if that is the same # of genes:
summary(ensembl_annot[ensembl_annot$type == "gene"])
# yep...

```

```{r shortest-gene-in-human-genome}
ensembl_df_genes <- ensembl_annot %>%
  as.data.frame %>%
  filter(type == "gene")

ensembl_df_genes %>%
  arrange(ensembl_df_genes$width) %>%
  head()
```

## What percentage of the genome are exons?

```{r}
# first make an object of exons and reduce them
total_exon_length <- ensembl_annot[ensembl_annot$type == "exon"] %>%
  # the reduce here takes the union of overlapping exons
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

total_exon_length/3.2e9 * 100
# Our filtered genome without the non autosomal chromosomes is about 5.7% exons

# Plot the range of exon sizes we can simply remove the sum function and plot
exon_lengths <- ensembl_annot[ensembl_annot$type == "exon"] %>%
  GenomicRanges::reduce()

# plotting
hist(width(exon_lengths), breaks = 100, xlim = c(0,10000))

# average exon length
average_exon_length <- mean(width(exon_lengths))
average_exon_length 
# it is 453nt
```

## What percentage of the genome is compriesed of gene bodies?

```{r}
# same thing as exons but change index in type to gene
reduced_gene_bodies <- ensembl_annot[ensembl_annot$type == "gene"] %>%
  GenomicRanges::reduce() %>%
  width() %>%
  sum()

reduced_gene_bodies/3.2e9 * 100
# ~70% of the genome is made up of gene bodies...

gene_body_lengths <- ensembl_annot[ensembl_annot$type == "gene"] %>%
  GenomicRanges::reduce()
  
hist(width(gene_body_lengths), breaks = 100, xlim = c(0,500000))

# What is average length of a gene body?

mean(width(gene_body_lengths))
# 38Kb

```
## Import Methylation files

```{r}
# Import Methylation loci DMLs and DMLs as GRanges -------------------------------
all_cpg_loci <- rtracklayer::import("HG38_02_07_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_CpG_Loci.bed")
genome(all_cpg_loci) <- "hg38"
all_dmls_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs.bed")
genome(all_dmls_gr) <- "hg38"
dmls_hyper_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs_HYPER.bed")
genome(dmls_hyper_gr) <- "hg38"
dmls_hypo_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMLs_HYPO.bed")
genome(dmls_hypo_gr) <- "hg38"
all_dmrs_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMRs.bed")
genome(all_dmrs_gr) <- "hg38"
dmrs_hyper_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMRs_HYPER.bed")
genome(dmrs_hyper_gr) <- "hg38"
dmrs_hypo_gr <- rtracklayer::import("HG38_24_06_2025_HCT116_3BKO_TIDY_VS_HCT116_WT_TIDY_DMRs_HYPO.bed")
genome(dmrs_hypo_gr) <- "hg38"
```


## Filter by gene and protein_coding (actually what I wanna do)
```{r manual_export_beds}

# Protein‐coding genes from GTF import 
genes_pc <- ensembl_annot[ensembl_annot$type == "gene" & ensembl_annot$gene_biotype == "protein_coding"]
gene_ids_pc <- genes_pc$gene_id

# Build each region by gene

## Promoters: 2kb upstream, 100bp downstream of the gene TSS ---------------
prom_manual <- unique(promoters(genes_pc, upstream   = 200, downstream = 200))
prom_manual <- trim(prom_manual)

## Immediate downstream: 1kb past the gene end
down_manual <- flank(genes_pc, width = 1000, start = FALSE)

## 5' and 3' UTRs (only protein‐coding) ---------------
utr5_manual <- ensembl_annot[ ensembl_annot$type == "five_prime_utr" & ensembl_annot$gene_biotype == "protein_coding" ]
utr3_manual <- ensembl_annot[ ensembl_annot$type == "three_prime_utr" & ensembl_annot$gene_biotype == "protein_coding" ]

## Exons & Introns (only protein‐coding genes) ---------------
exon_manual <- ensembl_annot[ ensembl_annot$type == "exon" & ensembl_annot$gene_biotype == "protein_coding" ]

intron_manual <- intronsByTranscript(ensembl_txdb, use.names=TRUE) %>%
  unlist() %>%
  subsetByOverlaps(genes_pc)  # keep only those from protein_coding genes

# Intergenic (everything else) ---------------
all_manual <- reduce(trim(c(prom_manual, down_manual, utr5_manual, utr3_manual, exon_manual, intron_manual
)))


chrom_sizes <- read.table("data/chrom_sizes/hg38.chr1_22.chrom.sizes", header = FALSE, col.names = c("chromosome", "length"))
genome <- GRanges(
  seqnames = chrom_sizes$chromosome,
  ranges = IRanges(start = rep(1, nrow(chrom_sizes)), end = chrom_sizes$length)
)

# Check the GRanges object
genome

# Assuming all_manual contains all your annotated regions (exons, promoters, UTRs, etc.)
overlaps <- findOverlaps(genome, all_manual)

# Identify the regions that do not overlap (these are the intergenic regions)
non_overlapping_indices <- setdiff(seq_along(genome), queryHits(overlaps))

# Extract the intergenic regions
intergenic_regions <- genome[non_overlapping_indices]

# Check the result
intergenic_regions

# rtracklayer::export(all_manual, "ALL_EXCEPT_INTERGENIC.BED", format = "bed")


# Put into a list and export each as BED -------------
region_list_manual <- GRangesList(
  Promoters           = prom_manual,
  immediateDownstream = down_manual,
  fiveUTRs            = utr5_manual,
  threeUTRs           = utr3_manual,
  Exons               = exon_manual,
  Introns             = intron_manual,
  Intergenic.Region   = intergenic_manual
)

# Export each to its own .bed for IGV
for (nm in names(region_list_manual)) {
  gr     <- region_list_manual[[nm]]
  mcols(gr) <- NULL               
  bedfile <- paste0(nm, "_manual.bed")
  rtracklayer::export(gr, bedfile, format = "bed")
  message("Wrote: ", bedfile)
}
```


## Understand how each of the chromosome regions are counted in ChipPeakAnno


## Calculate expected values

For each type of annotation (e.g. promoter regions, intragenic region etc ) check how many dmls do we expect to land on each of those by chance


## Perform fishers exact test for DMLs
